src/roi/Roi.ts

```diff
@@ -91,4 +138,584 @@ export class Roi {
+      ) {
+        let value = data[target - 1];
+        surroundingIDs.add(value);
+      }
+      if (
+        roiMap.width - column - roi.origin.column > 1 &&
+        data[target] === roi.id &&
+        data[target + 1] !== roi.id
+      ) {
+        let value = data[target + 1];
+        surroundingIDs.add(value);
+      }
+    }
+  }
+
+  // we check the first column and the last column
+  for (let column of [0, roi.width - 1]) {
+    for (let row = 0; row < roi.height; row++) {
+      let target = computeIndex(roi, row, column);
+      if (
+        row - roi.origin.row > 0 &&
+        data[target] === roi.id &&
+        data[target - roiMap.width] !== roi.id
+      ) {
+        let value = data[target - roiMap.width];
+        surroundingIDs.add(value);
+      }
+      if (
+        roiMap.height - row - roi.origin.row > 1 &&
+        data[target] === roi.id &&
+        data[target + roiMap.width] !== roi.id
+      ) {
+        let value = data[target + roiMap.width];
+        surroundingIDs.add(value);
+      }
+    }
+  }
+
+  return Array.from(surroundingIDs); // the selection takes the whole rectangle
+}
+
+/**
+ *
+ * @param roi - ROI
+ * @returns borders' length and their IDs
+ */
+function getBorders(roi: Roi): { ids: number[]; lengths: number[] } {
+  let roiMap = roi.getMap();
+  let data = roiMap.data;
+  let surroudingIDs = new Set<number>(); // allows to get a unique list without indexOf
+  let surroundingBorders = new Map();
+  let visitedData = new Set();
+  let dx = [+1, 0, -1, 0];
+  let dy = [0, +1, 0, -1];
+
+  for (
+    let column = roi.origin.column;
+    column <= roi.origin.column + roi.width;
+    column++
+  ) {
+    for (let row = roi.origin.row; row <= roi.origin.row + roi.height; row++) {
+      let target = column + row * roiMap.width;
+      if (data[target] === roi.id) {
+        for (let dir = 0; dir < 4; dir++) {
+          let newX = column + dx[dir];
+          let newY = row + dy[dir];
+          if (
+            newX >= 0 &&
+            newY >= 0 &&
+            newX < roiMap.width &&
+            newY < roiMap.height
+          ) {
+            let neighbour = newX + newY * roiMap.width;
+
+            if (data[neighbour] !== roi.id && !visitedData.has(neighbour)) {
+              visitedData.add(neighbour);
+              surroudingIDs.add(data[neighbour]);
+              let surroundingBorder = surroundingBorders.get(data[neighbour]);
+              if (!surroundingBorder) {
+                surroundingBorders.set(data[neighbour], 1);
+              } else {
+                surroundingBorders.set(data[neighbour], ++surroundingBorder);
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+  let ids: number[] = Array.from(surroudingIDs);
+  let borderLengths = ids.map((id) => {
+    return surroundingBorders.get(id);
+  });
+  return {
+    ids,
+    lengths: borderLengths,
+  };
+}
+function computeIndex(roi: Roi, y: number, x: number): number {
```

Comment:
I would add this method in the class so that we have access to 'this' and can simplify the code. This method can be public I think.
@stropitek WDYT ?